modules/apps/wiki/wiki-web/src/main/java/com/liferay/wiki/web/internal/importer/WikiImporterRegistry.java

```diff
@@ -57,23 +45,60 @@ public String getProperty(String importer, String key) {
 		return (String)serviceReference.getProperty(key);
 	}
 
-	public void modifiedService(
-		ServiceReference<WikiImporter> serviceReference) {
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTracker = new ServiceTracker<>(
+			bundleContext, WikiImporter.class,
+			new ServiceTrackerCustomizer<WikiImporter, WikiImporter>() {
 
-		removedService(serviceReference);
+				@Override
+				public WikiImporter addingService(
+					ServiceReference<WikiImporter> serviceReference) {
 
-		addingService(serviceReference);
-	}
+					String format = (String)serviceReference.getProperty(
+						"importer");
+
+					_serviceReferences.put(format, serviceReference);
+
+					return bundleContext.getService(serviceReference);
+				}
+
+				@Override
+				public void modifiedService(
```

Comment:
@JanisZhang I wonder if this is really needed. The only implementation of WikiImporter I find doesn't have any configuration, so it won't be "modified". Even if the implementation is a component and it does bind to a configuration, the modification won't cause a new instance to be created, and this registry doesn't get any info from the service.

I'm not sure about the behavior if someone registers an instance with `importer=A`, and then register another instance with same `importer=A`. Please either check the OSGi specification, or do some testing to confirm. But I don't think this would cause a "modified" either. If so, then it's unnecessary to replace the service when "modified".

Reply:
Hi @dantewang, I have done some tests and checked the OSGi specification. 

`If someone registers an instance with importer=A, and then register another instance with same importer=A.`
In this situation, the service won't be modified, just will adding another service.

The `modifiedService` method is called when the properties of the service are modified.

```
_serviceRegistration.setProperties( HashMapDictionaryBuilder.put(
                "importer", "modified Service"
        ).build());
```
then I think the original logic is unnecessary. If the properties are modified, it will updateProperties firstly then notify ServiceListeners, in this way the original `serviceReference` item in  `_serviceReferences` won't be removed since the current key is already set to new.
```
	public void modifiedService(
		ServiceReference<WikiImporter> serviceReference) {

		removedService(serviceReference);

		addingService(serviceReference);
	}

	public void removedService(
		ServiceReference<WikiImporter> serviceReference) {

		String importer = (String)serviceReference.getProperty("importer");

		_serviceReferences.remove(importer);
	}
```
What do you think?

Reply:
@JanisZhang You mean that when `importer` is modified to a new value, then it won't be possible for the service tracker to remove the previous entry (which has the previous key)?

If this is the case, I think you can just use ServiceTrackerMap instead. But actually, this class only needs the `importer` keys, which is even more weird to me... So I took a look, `WikiImporterRegistry` only has one usage, which is in a JSP. The render command puts the `WikiImporterRegistry` into the request, and the JSP uses it to get the keys (importers) to show on the page. The selected `importer` when you click some link/button on the page will then be passed to `WikiNodeLocalService.importPages` which has its own `ServiceTrackerMap`.

I thinks you can expose a method `WikiNodeLocalService.getImporterNames()`, and replace `WikiImporterRegistry` usage with it. In `ImportPagesMVCRenderCommand`, pass the `Collection` returned by that method, not the `WikiNodeLocalService` itself.

Reply:
Then you can remove `WikiImporterRegistry`.